File: HandSock-Server/src/main/java/com/server/handsock/service/TokenService.java
package com.server.handsock.service;

import com.server.handsock.console.MD5Encoder;
import org.springframework.stereotype.Service;
import com.server.handsock.console.TimeFormatUtils;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.UUID;
import java.util.concurrent.TimeUnit;

@Service
public class TokenService {

    private final RedisTemplate<String, String> redisTemplate;

    @Autowired
    public TokenService(RedisTemplate<String, String> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    public void removeUserToken(long uid) {
        String cachedToken = redisTemplate.opsForValue().get(String.valueOf(uid));
        if (cachedToken != null) redisTemplate.delete(String.valueOf(uid));
    }

    public String generateUserToken(long uid, String username, String address) {
        String formatTime = new TimeFormatUtils().formatTime("yyyy-MM-dd HH:mm:ss");
        String enCodeString = UUID.randomUUID().toString() + System.currentTimeMillis();
        String token = new MD5Encoder().toMD5(uid + username + enCodeString + formatTime + address);
        redisTemplate.opsForValue().set("handsock-userToken:" + uid, token, 30, TimeUnit.DAYS);
        return token;
    }

    public Boolean validUserToken(long uid, String oldToken) {
        String cachedToken = redisTemplate.opsForValue().get("handsock-userToken:" + uid);
        if (cachedToken != null)  {
            return cachedToken.equals(oldToken);
        } else return false;
    }
}

File: HandSock-Server/src/main/java/com/server/handsock/service/TokenService.kt
package com.server.handsock.service

import com.server.handsock.common.utils.HandUtils
import com.server.handsock.common.utils.JwtUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

@Service
class TokenService @Autowired constructor(private val redisTemplate: RedisTemplate<String, String>) {
    private fun removeUserToken(uid: Long) {
        val cachedToken = redisTemplate.opsForValue()["handsock-userToken:$uid"]
        if (cachedToken != null) redisTemplate.delete("handsock-userToken:$uid")
    }

    fun generateUserToken(uid: Long, username: String): String {
        val cachedToken = redisTemplate.opsForValue()["handsock-userToken:$uid"]
        if (cachedToken != null) return cachedToken // 直接使用原有Token（可实现多端同时在线）
        var token = JwtUtils.createToken(uid, username)
        token = HandUtils.encryptString(token)
        redisTemplate.opsForValue()["handsock-userToken:$uid", token, 30] = TimeUnit.DAYS
        return token
    }

    fun validUserToken(uid: Long, oldToken: String): Boolean {
        val redisKey = "handsock-userToken:$uid"
        return redisTemplate.opsForValue()[redisKey]?.let { cachedToken ->
            if (cachedToken != oldToken) return false
            try {
                val decryptedToken = HandUtils.decryptString(cachedToken)
                if (JwtUtils.verifyToken(decryptedToken)) {
                    true
                } else {
                    removeUserToken(uid)
                    false
                }
            } catch (e: Exception) {
                removeUserToken(uid)
                false
            }
        } ?: false
    }

    fun setScanStatus(qid: String, status: Int) {
        redisTemplate.opsForValue()["handsock-scanStatus:$qid", status.toString(), 30] = TimeUnit.SECONDS
    }

    fun removeScanStatus(qid: String) {
        val scanStatus = redisTemplate.opsForValue()["handsock-scanStatus:$qid"]
        if (scanStatus != null) redisTemplate.delete("handsock-scanStatus:$qid")
    }

    fun getScanStatus(qid: String): String? {
        val scanStatus = redisTemplate.opsForValue()["handsock-scanStatus:$qid"]
        return scanStatus
    }

    fun setScanTargetUser(qid: String, uid: String) {
        redisTemplate.opsForValue()["handsock-scanTargetUser:$qid", uid, 45] = TimeUnit.DAYS
    }

    fun getScanTargetUser(qid: String): String? {
        val targetUser = redisTemplate.opsForValue()["handsock-scanTargetUser:$qid"]
        return targetUser
    }

    fun setOpenApiCache(type: String, address: String) {
        redisTemplate.opsForValue()["handsock-openai-$type:$address", "ok", 3] = TimeUnit.DAYS
    }

    fun getOpenApiCache(type: String, address: String): Boolean {
        val requestStatus = redisTemplate.opsForValue()["handsock-openai-$type:$address"]
        return requestStatus == null
    }
}